diff -rupN a/sftp-server/log.c b/sftp-server/log.c
--- a/sftp-server/log.c	2011-12-07 17:28:18.760078091 +0400
+++ b/sftp-server/log.c	2011-12-07 17:28:18.904076649 +0400
@@ -428,3 +428,14 @@ do_log(LogLevel level, const char *fmt,
 	}
 	errno = saved_errno;
 }
+
+void
+fatal(const char *fmt,...)
+{
+	va_list args;
+
+	va_start(args, fmt);
+	do_log(SYSLOG_LEVEL_FATAL, fmt, args);
+	va_end(args);
+	exit(255);
+}
