diff -pru dropbear.orig/options.h dropbear/options.h
--- dropbear.orig/options.h	2012-05-02 11:44:10.919726329 +0400
+++ dropbear/options.h	2012-05-02 11:44:10.937726141 +0400
@@ -242,7 +242,7 @@ much traffic. */
 /* The default file to store the daemon's process ID, for shutdown
    scripts etc. This can be overridden with the -P flag */
 #ifndef DROPBEAR_PIDFILE
-#define DROPBEAR_PIDFILE "/data/dropbear/dropbear.pid"
+#define DROPBEAR_PIDFILE "/data/data/org.digimead.digi.ctrl.sshd/files/dropbear.pid"
 #endif
 
 /* The command to invoke for xauth when using X11 forwarding.
Только в dropbear: options.h.orig
