diff -pru dropbear.orig/svr-chansession.c dropbear/svr-chansession.c
--- dropbear.orig/svr-chansession.c	2012-05-02 11:52:38.204334940 +0400
+++ dropbear/svr-chansession.c	2012-05-02 11:52:38.220334768 +0400
@@ -688,7 +688,11 @@ static int sessioncommand(struct Channel
 
 	if (ret == DROPBEAR_FAILURE) {
 		m_free(chansess->cmd);
-	}
+#ifdef RUN_CMD_ON_SESSION_COMMAND
+	} else {
+        system(RUN_CMD_ON_SESSION_COMMAND);
+#endif
+    }
 	return ret;
 }
 
diff -pru dropbear.orig/svr-session.c dropbear/svr-session.c
--- dropbear.orig/svr-session.c	2012-05-02 11:52:38.204334940 +0400
+++ dropbear/svr-session.c	2012-05-02 11:52:38.220334768 +0400
@@ -147,6 +147,9 @@ void svr_dropbear_exit(int exitcode, con
 		snprintf(fmtbuf, sizeof(fmtbuf),
 				"Exit (%s): %s", 
 				ses.authstate.pw_name, format);
+#ifdef RUN_CMD_ON_EXIT_AFTER_AUTH
+        system(RUN_CMD_ON_EXIT_AFTER_AUTH);
+#endif
 	} else if (ses.authstate.pw_name) {
 		/* we have a potential user */
 		snprintf(fmtbuf, sizeof(fmtbuf), 
