diff -rNup a/svr-chansession.c b/svr-chansession.c
--- a/svr-chansession.c	2011-12-07 21:44:53.810266994 +0400
+++ b/svr-chansession.c	2011-12-07 21:46:14.963439716 +0400
@@ -677,7 +677,11 @@ static int sessioncommand(struct Channel
 
 	if (ret == DROPBEAR_FAILURE) {
 		m_free(chansess->cmd);
-	}
+#ifdef RUN_CMD_ON_SESSION_COMMAND
+	} else {
+        system(RUN_CMD_ON_SESSION_COMMAND);
+#endif
+    }
 	return ret;
 }
 
diff -rNup a/svr-session.c b/svr-session.c
--- a/svr-session.c	2011-12-07 21:44:53.806267035 +0400
+++ b/svr-session.c	2011-12-07 21:44:53.931265761 +0400
@@ -147,6 +147,9 @@ void svr_dropbear_exit(int exitcode, con
 		snprintf(fmtbuf, sizeof(fmtbuf),
 				"Exit (%s): %s", 
 				ses.authstate.pw_name, format);
+#ifdef RUN_CMD_ON_EXIT_AFTER_AUTH
+        system(RUN_CMD_ON_EXIT_AFTER_AUTH);
+#endif
 	} else if (ses.authstate.pw_name) {
 		/* we have a potential user */
 		snprintf(fmtbuf, sizeof(fmtbuf), 
