diff -rupN a/options.h b/options.h
--- a/options.h	2011-12-07 17:07:25.278767305 +0400
+++ b/options.h	2011-12-07 17:07:56.681449259 +0400
@@ -54,6 +54,9 @@ much traffic. */
 /* Enable X11 Forwarding - server only */
 #define ENABLE_X11FWD
 
+/* Check for valid user shell */
+/*#define ENABLE_VALID_SHELL_CHECK*/
+
 /* Enable TCP Fowarding */
 /* 'Local' is "-L" style (client listening port forwarded via server)
  * 'Remote' is "-R" style (server listening port forwarded via client) */
diff -rupN a/svr-auth.c b/svr-auth.c
--- a/svr-auth.c	2011-12-07 17:07:25.278767305 +0400
+++ b/svr-auth.c	2011-12-07 17:08:22.395188813 +0400
@@ -262,6 +262,7 @@ static int checkusername(unsigned char *
 		usershell = "/bin/sh";
 	}
 
+#ifdef ENABLE_VALID_SHELL_CHECK
 	/* check the shell is valid. If /etc/shells doesn't exist, getusershell()
 	 * should return some standard shells like "/bin/sh" and "/bin/csh" (this
 	 * is platform-specific) */
@@ -284,6 +285,7 @@ static int checkusername(unsigned char *
 goodshell:
 	endusershell();
 	TRACE(("matching shell"))
+#endif
 
 	TRACE(("uid = %d", ses.authstate.pw_uid))
 	TRACE(("leave checkusername"))
