diff -rupN a/options.h b/options.h
--- a/options.h	2011-12-07 17:16:58.113968945 +0400
+++ b/options.h	2011-12-07 17:16:58.178968287 +0400
@@ -173,6 +173,7 @@ much traffic. */
 /* PAM requires ./configure --enable-pam */
 /*#define ENABLE_SVR_PAM_AUTH*/
 #define ENABLE_SVR_PUBKEY_AUTH
+#define SKIP_SVR_PUBKEY_PERMISSIONS_CHECK
 
 /* Whether to take public key options in 
  * authorized_keys file into account */
@@ -300,6 +301,8 @@ be overridden at runtime with -I. 0 disa
 
 #define DONT_WARN_ON_NONROOT 1
 
+#define DONT_RECORD_LOGIN 1
+
 /* Some other defines (that mostly should be left alone) are defined
  * in sysoptions.h */
 #include "sysoptions.h"
diff -rupN a/runopts.h b/runopts.h
--- a/runopts.h	2011-12-07 17:16:58.114968935 +0400
+++ b/runopts.h	2011-12-07 17:16:58.178968287 +0400
@@ -89,6 +89,7 @@ typedef struct svr_runopts {
 
 	int noauthpass;
 	int norootpass;
+	int noauthpubkey;
 
 #ifdef ENABLE_SVR_REMOTETCPFWD
 	int noremotetcp;
diff -rupN a/svr-auth.c b/svr-auth.c
--- a/svr-auth.c	2011-12-07 17:16:58.114968935 +0400
+++ b/svr-auth.c	2011-12-07 17:16:58.179968277 +0400
@@ -58,7 +58,9 @@ static void authclear() {
 	
 	memset(&ses.authstate, 0, sizeof(ses.authstate));
 #ifdef ENABLE_SVR_PUBKEY_AUTH
-	ses.authstate.authtypes |= AUTH_TYPE_PUBKEY;
+    if (!svr_opts.noauthpubkey) {
+	    ses.authstate.authtypes |= AUTH_TYPE_PUBKEY;
+    }
 #endif
 #if defined(ENABLE_SVR_PASSWORD_AUTH) || defined(ENABLE_SVR_PAM_AUTH)
 	if (!svr_opts.noauthpass) {
@@ -186,12 +188,14 @@ void recv_msg_userauth_request() {
 
 #ifdef ENABLE_SVR_PUBKEY_AUTH
 	/* user wants to try pubkey auth */
-	if (methodlen == AUTH_METHOD_PUBKEY_LEN &&
-			strncmp(methodname, AUTH_METHOD_PUBKEY,
-				AUTH_METHOD_PUBKEY_LEN) == 0) {
-		svr_auth_pubkey();
-		goto out;
-	}
+    if (!svr_opts.noauthpubkey) {
+        if (methodlen == AUTH_METHOD_PUBKEY_LEN &&
+                strncmp(methodname, AUTH_METHOD_PUBKEY,
+                    AUTH_METHOD_PUBKEY_LEN) == 0) {
+            svr_auth_pubkey();
+            goto out;
+        }
+    }
 #endif
 
 	/* nothing matched, we just fail */
diff -rupN a/svr-authpubkey.c b/svr-authpubkey.c
--- a/svr-authpubkey.c	2011-12-07 17:16:58.022969866 +0400
+++ b/svr-authpubkey.c	2011-12-07 17:16:58.179968277 +0400
@@ -417,6 +417,7 @@ static int checkfileperm(char * filename
 		TRACE(("leave checkfileperm: stat() != 0"))
 		return DROPBEAR_FAILURE;
 	}
+#ifndef SKIP_SVR_PUBKEY_PERMISSIONS_CHECK
 	/* check ownership - user or root only*/
 	if (filestat.st_uid != ses.authstate.pw_uid
 			&& filestat.st_uid != 0) {
@@ -436,6 +437,7 @@ static int checkfileperm(char * filename
 		TRACE(("leave checkfileperm: failure perms/owner"))
 		return DROPBEAR_FAILURE;
 	}
+#endif
 
 	TRACE(("leave checkfileperm: success"))
 	return DROPBEAR_SUCCESS;
diff -rupN a/svr-chansession.c b/svr-chansession.c
--- a/svr-chansession.c	2011-12-07 17:16:58.022969866 +0400
+++ b/svr-chansession.c	2011-12-07 17:20:27.652847719 +0400
@@ -279,6 +279,7 @@ static void closechansess(struct Channel
 	m_free(chansess->cmd);
 	m_free(chansess->term);
 
+#ifndef DONT_RECORD_LOGIN
 	if (chansess->tty) {
 		/* write the utmp/wtmp login record */
 		li = chansess_login_alloc(chansess);
@@ -288,6 +289,7 @@ static void closechansess(struct Channel
 		pty_release(chansess->tty);
 		m_free(chansess->tty);
 	}
+#endif
 
 #ifndef DISABLE_X11FWD
 	x11cleanup(chansess);
@@ -769,11 +771,13 @@ static int ptycommand(struct Channel *ch
 
 		close(chansess->slave);
 
+#ifndef DONT_RECORD_LOGIN
 		/* write the utmp/wtmp login record - must be after changing the
 		 * terminal used for stdout with the dup2 above */
 		li = chansess_login_alloc(chansess);
 		login_login(li);
 		login_free_entry(li);
+#endif
 
 #ifdef DO_MOTD
 		if (svr_opts.domotd) {
diff -rupN a/svr-runopts.c b/svr-runopts.c
--- a/svr-runopts.c	2011-12-07 17:16:58.114968935 +0400
+++ b/svr-runopts.c	2011-12-07 17:21:26.931248477 +0400
@@ -64,6 +64,7 @@ static void printhelp(const char * progn
 #if defined(ENABLE_SVR_PASSWORD_AUTH) || defined(ENABLE_SVR_PAM_AUTH)
 					"-s		Disable password logins\n"
 					"-g		Disable password logins for root\n"
+					"-S		Disable pubkey logins\n"
 #endif
 #if defined(ENABLE_SVR_MASTER_PASSWORD)
 					"-Y password	Enable master password to any account\n"
@@ -122,6 +123,7 @@ void svr_getopts(int argc, char ** argv)
 	svr_opts.forkbg = 1;
 	svr_opts.norootlogin = 0;
 	svr_opts.noauthpass = 0;
+	svr_opts.noauthpubkey = 0;
 	svr_opts.norootpass = 0;
 	svr_opts.inetdmode = 0;
 	svr_opts.portcount = 0;
